src/ImageSharp/Image/ImageFrameCollection.cs

```diff
@@ -0,0 +1,166 @@
+﻿// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+using SixLabors.ImageSharp.PixelFormats;
+
+namespace SixLabors.ImageSharp
+{
+    /// <summary>
+    /// Encapsulates an imaged collection of frames.
+    /// </summary>
+    /// <typeparam name="TPixel">The type of the pixel.</typeparam>
+    public sealed class ImageFrameCollection<TPixel> : IEnumerable<ImageFrame<TPixel>>, IDisposable
+        where TPixel : struct, IPixel<TPixel>
+    {
+        private readonly IList<ImageFrame<TPixel>> frames = new List<ImageFrame<TPixel>>();
+
+        internal ImageFrameCollection(int width, int height)
+        {
+            this.Add(new ImageFrame<TPixel>(width, height));
+        }
+
+        internal ImageFrameCollection(IEnumerable<ImageFrame<TPixel>> frames)
+        {
+            Guard.NotNullOrEmpty(frames, nameof(frames));
+            foreach (ImageFrame<TPixel> f in frames)
+            {
+                this.Add(f);
+            }
+        }
+
+        /// <summary>
+        /// Gets the count.
+        /// </summary>
+        public int Count => this.frames.Count;
+
+        /// <summary>
+        /// Gets the root frame.
+        /// </summary>
+        public ImageFrame<TPixel> RootFrame => this.frames.Count > 0 ? this.frames[0] : null;
+
+        /// <summary>
+        /// Gets or sets the <see cref="ImageFrame{TPixel}"/> at the specified index.
+        /// </summary>
+        /// <value>
+        /// The <see cref="ImageFrame{TPixel}"/>.
+        /// </value>
+        /// <param name="index">The index.</param>
+        /// <returns>The <see cref="ImageFrame{TPixel}"/> at the specified index.</returns>
+        public ImageFrame<TPixel> this[int index]
+        {
+            get => this.frames[index];
+
+            set
+            {
+                this.ValidateFrame(value);
+                this.frames[index] = value;
+            }
+        }
+
+        /// <summary>
+        ///  Determines the index of a specific <paramref name="frame"/> in the <seealso cref="Image{TPixel}"/>.
+        /// </summary>
+        /// <param name="frame">The <seealso cref="ImageFrame{TPixel}"/> to locate in the <seealso cref="Image{TPixel}"/>.</param>
+        /// <returns>The index of item if found in the list; otherwise, -1.</returns>
+        public int IndexOf(ImageFrame<TPixel> frame) => this.frames.IndexOf(frame);
+
+        /// <summary>
+        ///  Inserts the <paramref name="frame"/> to the <seealso cref="Image{TPixel}"/> at the specified <paramref name="index"/>.
+        /// </summary>
+        /// <param name="index"> The zero-based index at which item should be inserted..</param>
+        /// <param name="frame">The <seealso cref="ImageFrame{TPixel}"/> to insert into the <seealso cref="Image{TPixel}"/>.</param>
+        public void Insert(int index, ImageFrame<TPixel> frame)
```

Comment:
Is it really a good idea to expose a mutable List API on `ImageFrameCollection`? Feels like the API surface is to wide now.
I'd rather keep it as an immuatble collection. If someone wants to implement a gif frame manipulator tool with ImageSharp (looks like a very uncommon use case), he should probably construct new image instances after calculating the frames.

Reply:
if its immutable then you can't create an animated gif.. also the public api doesn't allow users to specify frames so the only way for a user to go from zero to animated gif is to create a series of images, merge their frames together into a single image instance and then save that.

Reply:
It's probably matter of API taste, but I would utilize [builder pattern](https://en.wikipedia.org/wiki/Builder_pattern) instead, having some kind of `ImageBuilder` object to construct multi-frame images. This looks like a more convenient way to preserve valid object state. Constructing `ImageFrame` instances of a certain size, and verifying it against the image dimensions when calling `ImageFrameCollection.Add()` looks a bit clanky to me.

But looks like ImageSharp API follows mutable style on many other places (like `ImageMetaData`) so maybe it's easier to conform to this style everywhere for 1.0.

How about encapsulating the consistance rules into `ImageFrameCollection` by hiding `ImageFrame<T>` constructors and define methods on `ImageFrameCollection<T>`:
```C#
public ImageFrame<T> AppendNewFrame() { }
public ImageFrame<T> InsertNewFrame(int index) { }
```
?

@JimBobSquarePants thoughts?

Reply:
I had thought about adding those as APIs (well in concept not name, couldn't think of what to call them) but I like it.

This makes me think we should find a way to extend the `Mutate()` apis to allow us to apply mutations to single frames... might need to extend expand `IImageProcessors`, as we will need  `IFrameProcessor`s as well as `IImageProcessor`s will also have to scope some of the APIs to allow targeting only images and not frames (resize, crop, rotate etc) unless we allow images to have frames with multiple dimensions and handle normalising frame sizes on a per encoder bases instead of during add frame (i.e. update gif to draw based on the first frame dimensions)

Reply:
@antonfirsov I see where you're going there and it does seem safer. As far as I can see we don't actually new up a frame anywhere within our codebase.

@tocsoft I'd rather we treated images as a whole and avoid adding `IFrameProcessor`'s. Seems like we'd be creating a complexity level that would be difficult to follow.

Reply:
I think we should leave it as is for now then... we can worry about that later if/when people hit issues with the API.